File: app/api/v1/onboarding.py
"""Onboarding endpoints: goal-refinement gate and curriculum generation."""

import json
import uuid
from collections.abc import AsyncIterator
from typing import Any, Literal

from fastapi import APIRouter, HTTPException, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from app.api.deps import CurrentLearner, LLMClientDep, SessionDep
from app.services import onboarding, onboarding_sessions

router = APIRouter(tags=["onboarding"])


def _sse(obj: dict[str, Any]) -> str:
    """Format a dict as an SSE data frame."""
    return f"data: {json.dumps(obj)}\n\n"


class GoalTurnRequest(BaseModel):
    session_id: str
    content: str
    satisfied: bool = False
    mode: Literal["start", "resume"] = "start"


class GoalSessionResponse(BaseModel):
    """A server-issued onboarding session, owned by the learner who asked for it."""

    session_id: str


class CurriculumRequest(BaseModel):
    goal: str
    source_ids: list[uuid.UUID] | None = None


class CurriculumResponse(BaseModel):
    subject_name: str
    subject_description: str
    topics: list[dict]  # Topic dicts with 'name', 'description', 'kcs'


@router.post("/onboarding/goal-sessions", response_model=GoalSessionResponse)
async def start_goal_session(learner: CurrentLearner) -> GoalSessionResponse:
    """Mint the session id a goal-refinement negotiation runs under.

    The client used to invent this id, and the server keyed the negotiation's state on it with
    no learner attached — so knowing someone's id was enough to resume their onboarding.
    """
    return GoalSessionResponse(session_id=onboarding_sessions.issue(learner.id).session_id)


@router.post("/onboarding/goal-turns")
async def goal_refinement_turn(
    request: GoalTurnRequest,
    session: SessionDep,
    learner: CurrentLearner,
    llm: LLMClientDep,
) -> StreamingResponse:
    """Start or resume goal refinement, stream TurnEvent responses as SSE.

    Streams: token, awaiting_reply, committed, or error events.
    """

    try:
        onboarding_sessions.require(request.session_id, learner.id)
    except onboarding_sessions.NotYourSession as exc:
        # The same answer whether it belongs to someone else or never existed, so this cannot
        # be used to find out which ids are real.
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="onboarding session not found"
        ) from exc

    async def event_stream() -> AsyncIterator[str]:
        try:
            async for event in onboarding.run_goal_refinement_turn(
                llm=llm,
                session_id=request.session_id,
                user_content=request.content,
                satisfied=request.satisfied,
                resume=request.mode == "resume",
                learner_id=learner.id,
            ):
                if event.type == "token":
                    yield _sse({"type": "token", "text": event.text})
                elif event.type == "awaiting_reply":
                    yield _sse(
                        {"type": "awaiting_reply", "text": event.text, "detail": event.detail}
                    )
                elif event.type == "committed":
                    yield _sse({"type": "committed", "goal": event.text, "detail": event.detail})
                elif event.type == "error":
                    yield _sse({"type": "error", "detail": event.detail})
        except Exception as exc:
            yield _sse({"type": "error", "detail": str(exc)})

    return StreamingResponse(event_stream(), media_type="text/event-stream")


@router.post("/onboarding/curriculum", response_model=CurriculumResponse)
async def generate_curriculum_endpoint(
    request: CurriculumRequest,
    session: SessionDep,
    learner: CurrentLearner,
    llm: LLMClientDep,
) -> CurriculumResponse:
    """Generate a curriculum proposal from goal and optional source excerpts.

    Returns: CurriculumResponse or 400 on LLM failure.
    """
    proposal = await onboarding.generate_curriculum_for_onboarding(
        session=session,
        llm=llm,
        goal=request.goal,
        source_ids=request.source_ids,
        learner_id=learner.id,
    )
    if proposal is None:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Curriculum generation failed. Please try again.",
        )

    # Convert CurriculumProposal to response format
    topics = []
    for topic in proposal.topics:
        topic_dict = {
            "name": topic.name,
            "description": topic.description,
            "kcs": [{"name": kc.name, "description": kc.description} for kc in topic.kcs],
        }
        topics.append(topic_dict)

    return CurriculumResponse(
        subject_name=proposal.subject_name,
        subject_description=proposal.subject_description,
        topics=topics,
    )

File: frontend/src/api/onboarding.ts
import { useCallback, useRef, useState } from "react";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiFetch } from "./client";

// ============================================================================
// Types
// ============================================================================

export type OnboardingTurnEvent =
  | { type: "token"; text: string }
  | { type: "awaiting_reply"; text: string; detail: string }
  | { type: "committed"; goal: string; detail: string }
  | { type: "error"; detail: string };

export interface KCProposal {
  name: string;
  description: string;
  /** Stable handle for this KC within one proposal, assigned by the backend's parser. */
  key?: string;
  /** Keys of the KCs that come first — the prerequisite edges the planner orders by (S22).
   * Declared so the review step's edits carry them: they survive today only because every
   * handler spreads the object, and a refactor that rebuilt one would drop them silently.
   * Renaming a KC is safe (edges reference keys, not names), and deleting one leaves a
   * dangling reference that the backend drops rather than failing the commit. */
  requires?: string[];
}

export interface TopicProposal {
  name: string;
  description: string;
  kcs: KCProposal[];
}

export interface CurriculumProposal {
  subject_name: string;
  subject_description: string;
  topics: TopicProposal[];
}

export interface GoalTurnBody {
  session_id: string;
  content: string;
  satisfied: boolean;
  mode: "start" | "resume";
}

export interface SubjectCommitPayload {
  subject_name: string;
  subject_description: string | null;
  topics: TopicProposal[];
  source_ids: string[] | null;
}

export interface CreatedSubject {
  id: string;
  slug: string;
  name: string;
  description: string | null;
}

// ============================================================================
// SSE Streaming
// ============================================================================

/**
 * Asks the server for a goal-refinement session id.
 *
 * The client used to invent this id, which meant the server keyed a negotiation's state on a
 * value it had never issued and could not attribute to anyone.
 */
export async function createGoalSession(): Promise<string> {
  const res = await apiFetch(`/api/v1/onboarding/goal-sessions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
  });
  if (!res.ok) throw new Error(`could not start a goal session: ${res.status}`);
  return ((await res.json()) as { session_id: string }).session_id;
}

/**
 * Streams one goal-refinement turn. Mirrors sse.ts's streamTurn pattern: POST returns an SSE
 * body, so we read the response body's ReadableStream directly and parse `data: {...}\n\n`
 * frames by hand.
 */
export async function* streamGoalTurn(
  body: GoalTurnBody,
  signal?: AbortSignal,
): AsyncGenerator<OnboardingTurnEvent> {
  const res = await apiFetch(`/api/v1/onboarding/goal-turns`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
    signal,
  });
  if (!res.ok || !res.body) {
    throw new Error(`stream failed: ${res.status} ${res.statusText}`);
  }

  const reader = res.body.pipeThrough(new TextDecoderStream()).getReader();
  let buffer = "";
  for (;;) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += value;
    const frames = buffer.split("\n\n");
    buffer = frames.pop() ?? "";
    for (const frame of frames) {
      const line = frame.split("\n").find((l) => l.startsWith("data: "));
      if (!line) continue;
      yield JSON.parse(line.slice("data: ".length)) as OnboardingTurnEvent;
    }
  }
}

// ============================================================================
// Hooks
// ============================================================================

/**
 * The server-issued session id this wizard's negotiation runs under. Minted once and kept for
 * the life of the component; a new wizard is a new negotiation.
 */
export function useGoalSession() {
  return useQuery({
    queryKey: ["goal-session"],
    queryFn: createGoalSession,
    staleTime: Infinity,
    gcTime: Infinity,
    refetchOnMount: false,
    refetchOnWindowFocus: false,
  });
}

/**
 * Owns the goal-refinement negotiation: local streaming state, the latest proposal text,
 * the committed goal (once accepted), and a send function that streams turns.
 *
 * Tracks whether this session has already started (to determine mode: "start" vs "resume").
 */
export function useGoalRefinement(sessionId: string | undefined) {
  const [proposal, setProposal] = useState<string>("");
  const [committedGoal, setCommittedGoal] = useState<string | null>(null);
  const [isStreaming, setIsStreaming] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [tokens, setTokens] = useState<string>("");
  const hasStartedRef = useRef(false);

  const send = useCallback(
    async (content: string, satisfied: boolean) => {
      if (!sessionId || isStreaming) return;
      setError(null);
      setTokens("");
      setIsStreaming(true);

      const mode: "start" | "resume" = hasStartedRef.current ? "resume" : "start";
      if (!hasStartedRef.current) {
        hasStartedRef.current = true;
      }

      try {
        for await (const ev of streamGoalTurn({
          session_id: sessionId,
          content,
          satisfied,
          mode,
        })) {
          if (ev.type === "token") {
            setTokens((t) => t + ev.text);
          } else if (ev.type === "awaiting_reply") {
            setProposal(ev.text);
            setTokens("");
          } else if (ev.type === "committed") {
            setCommittedGoal(ev.goal);
            setTokens("");
          } else if (ev.type === "error") {
            setError(ev.detail);
          }
        }
      } catch (e) {
        setError(e instanceof Error ? e.message : "Goal refinement failed.");
      } finally {
        setIsStreaming(false);
      }
    },
    [sessionId, isStreaming],
  );

  return { proposal, committedGoal, isStreaming, error, tokens, send };
}

/**
 * Mutation for generating a curriculum proposal from a committed goal and optional sources.
 */
export function useGenerateCurriculum() {
  return useMutation({
    mutationFn: async ({
      goal,
      sourceIds,
    }: {
      goal: string;
      sourceIds: string[] | null;
    }): Promise<CurriculumProposal> => {
      const res = await apiFetch(`/api/v1/onboarding/curriculum`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ goal, source_ids: sourceIds }),
      });
      if (!res.ok) {
        throw new Error(`curriculum generation failed: ${res.status} ${res.statusText}`);
      }
      return res.json();
    },
  });
}

/**
 * Mutation for committing a subject (with its knowledge graph structure) to the learner's
 * subjects. Invalidates the subjects query on success.
 */
export function useCommitSubject() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (payload: SubjectCommitPayload): Promise<CreatedSubject> => {
      const res = await apiFetch(`/api/v1/subjects/commit`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
      if (!res.ok) {
        const error = new Error(
          `subject commit failed: ${res.status} ${res.statusText}`,
        ) as Error & { status?: number };
        error.status = res.status;
        throw error;
      }
      return res.json();
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["subjects"] });
    },
  });
}
